crates/matrix-sdk-common/src/linked_chunk/builder.rs

```diff
@@ -0,0 +1,448 @@
+            // looping.
+            let cur_temp = self
+                .chunks
+                .remove(&cur_id)
+                .ok_or(LinkedChunkBuilderError::MissingChunk { id: cur_id })?;
+
+            // Graduate the current temporary chunk into a final chunk.
+            let mut chunk_ptr = match cur_temp.content {
+                ChunkContent::Gap(gap) => Chunk::new_gap_leaked(cur_id, gap),
+
+                ChunkContent::Items(items) => {
+                    // Update the total count of items.
+                    num_items += items.len();
+                    Chunk::new_leaked(cur_id, ChunkContent::Items(items))
+                }
+            };
+
+            if first_chunk_ptr.is_none() {
+                first_chunk_ptr = Some(chunk_ptr);
+            }
+
+            // Link the current chunk to its previous one, and its previous's next to it.
+            let chunk = unsafe { chunk_ptr.as_mut() };
+
+            let prev_id = if let Some(mut prev_chunk_ptr) = prev_chunk_ptr {
+                let prev_chunk = unsafe { prev_chunk_ptr.as_mut() };
+                prev_chunk.next = Some(chunk_ptr);
+                Some(prev_chunk.identifier())
+            } else {
+                None
+            };
+
+            chunk.previous = prev_chunk_ptr;
+
+            // The declared previous link was never checked, if the chunk wasn't part of the
+            // chase of the first chunk (in the first loop in this function). In
+            // theory this check isn't necessary, because another future
+            // iteration would fail with one of the other error variants, since
+            // we're removing chunks from `self.chunks`; this makes errors easier
+            // to understand, though.
+            if cur_temp.previous != prev_id {
+                return Err(LinkedChunkBuilderError::InconsistentLinks {
+                    chunk: cur_id,
+                    declared_prev: cur_temp.previous,
+                    actual_prev: prev_id,
+                });
+            }
+
+            // Update the maximum chunk identifier, while we're around.
+            max_chunk_id = max_chunk_id.max(cur_id.index());
+
+            // Prepare for next iteration.
+            prev_chunk_ptr = Some(chunk_ptr);
+            maybe_cur_id = cur_temp.next;
+        }
+
+        // At this point, we should have gone through all the chunks in `self.chunks`,
+        // otherwise we had multiple connected components in the first place,
+        // and this wasn't a linked list.
+        if !self.chunks.is_empty() {
+            let remaining = Vec::from_iter(self.chunks.into_keys());
+            return Err(LinkedChunkBuilderError::MultipleConnectedComponents { chunks: remaining });
+        }
+
+        // SAFETY: if one of these unwrap fails, it means we had no chunks in the stash,
+        // so we should have returned early.
+        let first = first_chunk_ptr.unwrap();
```

Comment:
Is there really no other way to express this without unwrapping?

In any case please use an `except()` if there isn't, unwraps are only ok in tests or if you're using an `std` `Mutex`.

Reply:
with a bit extra effort it's possible to get rid of the two options for first and last :+1: